cub/block/block_run_length_decode.cuh

```diff
@@ -0,0 +1,435 @@
+   */
+  template <typename DecodeBufferOffsetT>
+  __device__ __forceinline__ void WriteBeginningOfRun(UnzippedRunLengthDecodeBuffer &decode_buffer,
+                                                      DecodeBufferOffsetT buffer_offset,
+                                                      const UniqueItemT &unique_item,
+                                                      const RelativeRunOffsetT &relative_offset)
+  {
+    decode_buffer.uniques[buffer_offset] = unique_item;
+    decode_buffer.offsets[buffer_offset] = relative_offset;
+  }
+
+  /**
+   * [SPECIALIZATION: NORMAL OR OFFSETS]
+   * Populates the run-length decode buffer with <continuation-of-current-run> items
+   */
+  __device__ __forceinline__ void LoadFromDecodeBuffer(ZippedRunLengthDecode (&scan_items)[DECODED_ITEMS_PER_THREAD],
+                                                       UnzippedRunLengthDecodeBuffer &decode_buffer)
+  {
+    RelativeOffsetT run_offset = linear_tid * DECODED_ITEMS_PER_THREAD;
+#pragma unroll
+    for (uint32_t i = 0; i < DECODED_ITEMS_PER_THREAD; i++)
+    {
+      scan_items[i].unique = decode_buffer.uniques[run_offset];
+      scan_items[i].offset = decode_buffer.offsets[run_offset];
+      run_offset++;
+    }
+  }
+
+  /**
+   * [SPECIALIZATION: NORMAL OR OFFSETS]
+   * Populate registers (SOA) to return the results to the user
+   */
+  template <typename UserRelativeOffsetT>
+  __device__ __forceinline__ void PopulateReturnValues(ZippedRunLengthDecode (&scan_items)[DECODED_ITEMS_PER_THREAD],
+                                                       UniqueItemT (&uniques)[DECODED_ITEMS_PER_THREAD],
+                                                       UserRelativeOffsetT (&relative_offset)[DECODED_ITEMS_PER_THREAD])
+  {
+#pragma unroll
+    for (uint32_t i = 0; i < DECODED_ITEMS_PER_THREAD; i++)
+    {
+      uniques[i]         = scan_items[i].unique;
+      relative_offset[i] = scan_items[i].offset;
+    }
+  }
+
+public:
+  struct TempStorage : Uninitialized<_TempStorage>
+  {};
+
+  //---------------------------------------------------------------------
+  // CONSTRUCTOR
+  //---------------------------------------------------------------------
+  __device__ __forceinline__ BlockRunLengthDecode()
+      : temp_storage(PrivateStorage())
+      , linear_tid(RowMajorTid(BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z))
+  {}
+
+  __device__ __forceinline__ BlockRunLengthDecode(TempStorage &temp_storage)
+      : temp_storage(temp_storage.Alias())
+      , linear_tid(RowMajorTid(BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z))
+  {}
+
+  /**
+   * @brief Initializes the run-length decoding instances with the given sequence of runs and returns the total
+   * run-length decoded size. Subsequent calls to <b>RunLengthDecode</b> can be used to retrieve the run-length
+   * decoded data.
+   */
+  template <typename RunLengthT, typename OffsetT>
+  __device__ __forceinline__ void Init(UniqueItemT (&unique_items)[RUNS_PER_THREAD],
```

Comment:
This method looks like a part of the constructor. We can't use other methods until this one is called.  I understand that this way, we can reuse the `BlockRunLengthDecode` object. But its non-Init constructor part is pretty lightweight, so there should be no performance issue in constructing a new one each time. Can we make this method private and call it from the constructor, or did I miss something?

Reply:
I think we can use something like [this](https://godbolt.org/z/r7TGErqG6). Or we can expose the PrefixProviderT as a template parameter and receive it in constructor :sweat_smile:

Reply:
Anyway, I consider your idea of exposing an API that allows clients to provide us with an already computed prefix sum a must.

Reply:
Have added the interface which allows the user to provide either (a) the array of the runs' lengths or (b) the array of the beginning-offsets of each run's length.